Add cli.py that parses arguments into a config object

tmillenaar opened the pull request:
Based on https://github.com/transientskp/pyse/pull/127

TODO:

 - [x] Parse command line arguments
 - [x] Make sure all relevant arguments are represented in config toml and CLI parameters
 - [x] Read config toml and update with parameters found on command line
 - [x] ~~Create run function that calls PySE with supplied arguments~~ -> _Handled by existing pyse.py_
 - [x] ~~Write data to file~~ -> _Handled by existing pyse.py_

Run pyse through command line like so

`pyse test/data/GRB120422A-120429.fits --csv --config-file test/data/config.toml --detection 6 --analysis 5`

If you want to use hatch, use `hatch run default:pyse test/data/GRB120422A-120429.fits --csv --config_file test/data/config.toml --detection 6 --analysis 5`

HannoSpreeuw commented:
Wrt the second bullet, let me try to list all relevant parameters, starting with the command line.

Command-line parameters to process the image (required representation in [ImgConf](https://github.com/transientskp/pyse/blob/f6eb6e83161963612b5f49693b5324645a6125fc/sourcefinder/config.py#L135)) from [pyse.py](https://github.com/transientskp/pyse/blob/master/scripts/pyse.py):

- [ ] detection
- [ ] analysis
- [ ] fdr
- [ ] alpha
- [ ] deblend-thresholds
- [ ] grid
- [ ] margin
- [ ] radius
- [ ] bmaj
- [ ] bmin
- [ ] bpa
- [ ] force-beam
- [ ] detection-image
- [ ] fixed-posns
- [ ] fixed-posns-file
- [ ] ffbox

Command-line parameters for requested output products (required representation in [ExportSettings](https://github.com/transientskp/pyse/blob/f6eb6e83161963612b5f49693b5324645a6125fc/sourcefinder/config.py#L173C7-L173C21)) from [pyse.py](https://github.com/transientskp/pyse/blob/master/scripts/pyse.py):

- [ ] skymodel
- [ ] csv
- [ ] regions
- [ ] rmsmap
- [ ] sigmap
- [ ] residuals
- [ ] islands

Ticking the boxes for representation in cli.py.

HannoSpreeuw commented on sourcefinder/utility/cli.py:
Should it not be a `git mv` as in 
`git mv sourcefinder/utility/cli.py sourcefinder/utility/monitoring.py`?

tmillenaar replied:
I decided to keep and expand on sourcefinder/utility/cli.py instead adding a second cli.py

tmillenaar commented:
I now tried to update the existing scripts/pyse.py and utility/cli.py instead of creating my own entry-point.
I encountered a bunch of cli arguments that are required to make pyse.py run that were not in the proposed ImgConf and ExportConf settings. I put those here for now:
https://github.com/transientskp/pyse/blob/b791dfa5ca115dcdbee8489d82b46586110bc328/sourcefinder/config.py#L184

@HannoSpreeuw How would you propose we handle these? After reading your previous comment in this thread (https://github.com/transientskp/pyse/pull/134#issuecomment-2886940398) it sounds like you want these added to ImgConf. Should I append them to the existing one here: https://github.com/transientskp/pyse/blob/b791dfa5ca115dcdbee8489d82b46586110bc328/sourcefinder/config.py#L136 instead of the separate ExtractionSettings I linked before?

Also, you mentioned that islands and residuals are purely export flags, but I found they are also referenced when processing the image here:
https://github.com/transientskp/pyse/blob/b791dfa5ca115dcdbee8489d82b46586110bc328/sourcefinder/image.py#L1145
https://github.com/transientskp/pyse/blob/b791dfa5ca115dcdbee8489d82b46586110bc328/sourcefinder/image.py#L1149
Unless we rework that bit of image.py, we do need these parameters in ImgConf

HannoSpreeuw commented:
> After reading your previous comment in this thread ([#134 (comment)](https://github.com/transientskp/pyse/pull/134#issuecomment-2886940398)) it sounds like you want these added to ImgConf. Should I append them to the existing one here:
> 
> https://github.com/transientskp/pyse/blob/b791dfa5ca115dcdbee8489d82b46586110bc328/sourcefinder/config.py#L136

Yes, please add them there instead of introducing a separate _ExtractionSettings_ class.

We should probably add a docstring to both the _ImgConf_ as well as the _ExportSettings_ class, for clarfication:
 _ImgConf_ should cover all the specifications for processing the image.
_ExportSettings_  should cover all the desired export products.

_islands_ and _residuals_ are indeed purely export flags.

They do appear in the _image_ module, but that does not mean that these parameters have any effect on the way the image is processed.
They are just boolean flags, i.e. yes, I want this, or no, I don't want this to be exported.

They are needed in the _image_ module because that is, logically, the most suitable module to trigger the computation of the residuals and islands maps.

**It does imply that we need  _Conf_ - instead of only _ImgConf_ - in the _image_ module.**

HannoSpreeuw commented on sourcefinder/config.py:
```diff
@@ -153,46 +150,47 @@ class ImgConf(_Validate):
-    residuals: bool = False
-    islands: bool = False
-    source_params: list[str] = field(default_factory=lambda: _source_params)
+    """Configuration for writing out the image data."""
```
Perhaps writing out "image data" is not specific enough.
Since "image data" could refer to the observational image data, i.e. the input.
Something like "Selection of output, related to detected sources and/or intermediate image processing products." may be clearer.
The noise map (rmsmap) is an example of an intermediate image processing product.

tmillenaar replied:
Sure thing, changed it to your suggestion

HannoSpreeuw commented on test/test_scripts/test_pyse.py:
It's great that the pyse.py script is now covered by a unit test!

tmillenaar commented:
In the past few days I changed the following:

 - Get rid of the ExtractionSettings I introduced and merge it with the already existing ImgConf
 - Put Islands and Residuals from ImgConf to ExportSettings
 - Reference full Conf in image.py instead of only ImgConf so we can still refer to Islands and Residuals, now on conf.export
 - Add an optional 'output_dir' cli argument (required for testing with temporary directories but likely handy regardless)
 - Add some tests for pyse.py, mostly focussed on checkin if the specified export products are present
 - Remove default values fom cli.py, these are defined in the config.toml which would otherwise always be overwritten by the argparse default


I think with that I tackled all open items in this MR. I will make a new MR for handling source_params and serialize mentioned here: https://github.com/transientskp/pyse/pull/127#issuecomment-2873589035

@HannoSpreeuw Is this MR close enough to the implementation you had in mind? I would not be surprised if I missed something or misinterpreted an earlier comment ;)

HannoSpreeuw commented:
Great job.

The only thing I can think of at this point, would be to add the desired source parameters to ExportSettings.
At this point we have these hard coded in [csv](https://github.com/transientskp/pyse/blob/0622a50edefa62fc2d8b61beb48bcb1e2b7253fb/scripts/pyse.py#L89):
```
    for source in sourcelist:
        values = (
            source.ra, source.ra.error, source.dec, source.dec.error,
            source.smaj_asec, source.smaj_asec.error, source.smin_asec,
            source.smin_asec.error, source.theta_celes,
            source.theta_celes.error, source.flux, source.flux.error,
            source.peak, source.peak.error, source.x, source.y, source.sig,
            source.reduced_chisq)
```
while we would like `values` to be configurable.            

I just added something similar to MR #127 , fourth bullet, i.e. we want the source parameters to be exported to TraP configurable.

And a handle to output to hdf5 would be nice, that's #129 
But that can be added later, I guess.

@suvayu Any other comments?

suvayu reviewed (changes requested):
One general comment: I really like the use of `pdb` to debug on crash, very cool!

PS: I hope I got everything.

tmillenaar commented on pyproject.toml:
```diff
@@ -29,6 +29,9 @@ dependencies = [
 Homepage = "http://docs.transientskp.org/"
 
+[project.scripts]
+pyse = "scripts.pyse:main"
```
Yeah good catch, the scripts folder is indeed not installed. It only worked for me because I ran it from the cloned repository. Based on your feedback I deprecated scripts/pyse.py, moved the remaining functions from there to sourcefinder/utility/cli.py and made that the entry point for the pyse command as defined in the pyproject.toml.
Is this what you had in mind?

suvayu replied:
Yup!

tmillenaar commented on sourcefinder/utility/cli.py:
```diff
@@ -28,3 +34,240 @@ def parse_monitoringlist_positions(args, str_name="monitor_coords",
+    if not_visited:
+        raise ValueError(f"The defaults for the following CLI arguments are missing in the config {not_visited}")
+
+    return Conf(**combined_data)
```
Alright, indeed I had to do some book-keeping because argparse does not have the ability build in to preserve the group structure on parse. I was at least able to prevent having to define the structure manually, by looping over the internal argparse objects and checking their group, then organizing the arguments based on that.

This new approach retires the recursive update function, as well as the _read_conf_as_dict function. I assume the new approach matches your preference a bit more in regards to the use of dataclasses. Let me know if this is indeed more what you had in mind. See the following diff: https://github.com/transientskp/pyse/pull/134/commits/da3eccaf7fb7fde5aa6f149e1540964bb84fe7b0

tmillenaar commented on sourcefinder/utility/cli.py:
```diff
@@ -28,3 +34,240 @@ def parse_monitoringlist_positions(args, str_name="monitor_coords",
+    # Finally, positional arguments- the file list:
+    parser.add_argument('files', nargs='+',
+                        help="Image files for processing")
+    return parser
```
Can you elaborate a bit? Why do you want them to be mutually exclusive? Also, are the flags not already in the `--my-flag` format?

suvayu replied:
I was hoping the mutually exclusive option would lead to disentangling the "mode" issue.  But I agree with your comment there, let's handle this separately.

As for the flags, many flags are underscored, e.g. `--rms_filter`, `--back_size_{x,y}`, and many others.  These should be hyphenated, as that's more natural on the CLI.  `argparse` handles this normally:
```python
In [2]: parser = ArgumentParser()

In [3]: parser.add_argument("--foo-bar", action="store_true")

In [4]: parser.parse_args(["--foo-bar"])
Out[4]: Namespace(foo_bar=True)
```

tmillenaar commented on scripts/pyse.py:
```diff
@@ -220,22 +176,22 @@ def handle_args(args=None):
+        mode = "detimage"  # mode 1.2 above
     else:
-        options.mode = "threshold"  # mode 1.1 above
+        mode = "threshold"  # mode 1.1 above
```
I am hesitant about changing this, because it is a non-trivial amount of work without a clear upside. I realize the mode as was implemented already is not very ergonomic, but it is quite entangled. The relation between these arguments is conditional so in run_sourcefinder we cannot do a simple "if conf.image.detection_image" and follow with the logic. We would also have to check for example if fixed_coords is not set as well as the other related arguments, because they would override the detection_image setting as it is currently implemented. Reworking how this mode is used requires rethinking the relations between these arguments and a substantial refactor.

If it is a part of the code that has bothered you for a while I propose to follow this MR up with a new one dedicated to that rework. Agreed?

suvayu replied:
Hmm, you are right, let's tackle this in a separate PR.

suvayu commented:
@tmillenaar I left some comments, basically I will consider this mergeable after you address the underscore in flags issue: `--my-flag` in favour of `--my_flag`.  On the other comments, agree with you, it's best to handle separately.

tmillenaar commented:
> > This was surprisingly more involved than I imagined it would be, but I think I've got it now. Please double check if the default params are set how you'd like them. It is in this branch now but to check the diff of the commit, see: [fb02478](https://github.com/transientskp/pyse/commit/fb024785a18483df41335e7562237b91220feeca)
> 
> Thanks! It seems you have also addressed [part of PR 127](https://github.com/transientskp/pyse/pull/127#issuecomment-2873589035), much appreciated.
> 
> This MR is evolving into a major effort, but together we'll get there.
> 
> These are my comments:
> 
>     1. Could you please add [ew_sys_err](https://github.com/transientskp/pyse/blob/fb024785a18483df41335e7562237b91220feeca/sourcefinder/extract.py#L1160) and [ns_sys_err](https://github.com/transientskp/pyse/blob/fb024785a18483df41335e7562237b91220feeca/sourcefinder/extract.py#L1161C9-L1161C19) to [ImgConf](https://github.com/transientskp/pyse/blob/fb024785a18483df41335e7562237b91220feeca/sourcefinder/config.py#L139)? Sorry, I should have specified that earlier.
> 
>     2. The [default set of parameters to be exported to a csv file](https://github.com/transientskp/pyse/blob/a43b64d684775605051adf9f754bb0ce6eda3493/scripts/pyse.py#L94) should be retained instead of copied from the [default set of parameters to be passed on to TraP](https://github.com/transientskp/pyse/blob/fb024785a18483df41335e7562237b91220feeca/sourcefinder/config.py#L116). Perhaps this is taken care of, but at a first glance these sets seemed not to have been defined independently.
> 
>     3. Some PEP 8 compliance, so [this line](https://github.com/transientskp/pyse/blob/fb024785a18483df41335e7562237b91220feeca/sourcefinder/extract.py#L1201) and [this line](https://github.com/transientskp/pyse/blob/fb024785a18483df41335e7562237b91220feeca/sourcefinder/extract.py#L1206) are slightly long.
> 
> 
> Other than that, ready to merge.

1. Yeah  the xx_sys_err values were a bit of a nuisance to me as well. Now it is in ImgConf that cleans up a bit.
2. In the new setup the default csv parameters are taken from the config.toml: https://github.com/transientskp/pyse/blob/fb024785a18483df41335e7562237b91220feeca/test/data/config.toml#L41 The defaults in the class (e.g. https://github.com/transientskp/pyse/blob/fb024785a18483df41335e7562237b91220feeca/sourcefinder/config.py#L195) are overwritten by the values found in the config.toml and the cli arguments. If we want the default CSV columns to remain the same I think we should make that change in the config.toml, agreed?
3. Sure thing, handled

HannoSpreeuw commented:
> Oh I though the whole purpose of source_params in the config.toml was such that a user can specify what attributes to export (to the csv).

That's correct.

> Writing a csv is only used when calling pyse as a standalone package at which point the config.toml and command line arguments are used to generate a config.

But in case one does not provide a config.toml, there should be some default settings to prevent PySE from crashing.
And these default settings should come from the config module.

Of course CLI will override config.toml and config.toml will override the default settings in the config module, following [this sequence](https://github.com/transientskp/pyse/pull/127#issuecomment-2873484276).

HannoSpreeuw commented:
Nice, almost ready to merge.
Somehow `hatch run test:pytest` gives
```
8 failed, 209 passed
```
These fails all seem to have the same cause:
```
FAILED test/test_scripts/test_pyse.py::test_pyse_simple - FileNotFoundError: [Errno 2] No such file or directory: 'pyse'
FAILED test/test_scripts/test_pyse.py::test_pyse_export[--skymodel] - FileNotFoundError: [Errno 2] No such file or directory: 'pyse'
FAILED test/test_scripts/test_pyse.py::test_pyse_export[--csv] - FileNotFoundError: [Errno 2] No such file or directory: 'pyse'
FAILED test/test_scripts/test_pyse.py::test_pyse_export[--regions] - FileNotFoundError: [Errno 2] No such file or directory: 'pyse'
FAILED test/test_scripts/test_pyse.py::test_pyse_export[--rmsmap] - FileNotFoundError: [Errno 2] No such file or directory: 'pyse'
FAILED test/test_scripts/test_pyse.py::test_pyse_export[--sigmap] - FileNotFoundError: [Errno 2] No such file or directory: 'pyse'
FAILED test/test_scripts/test_pyse.py::test_pyse_export[--residuals] - FileNotFoundError: [Errno 2] No such file or directory: 'pyse'
FAILED test/test_scripts/test_pyse.py::test_pyse_export[--islands] - FileNotFoundError: [Errno 2] No such file or directory: 'pyse'
```

HannoSpreeuw commented:
> Since I don't plan on using the pre-defined source_params from TraP, (I will supply a Conf object where I define my own columns)

Ah. Interesting comment, which immediately shows the enhanced flexibility from the Conf object.
And you must have adapted the unit tests, involving the `extract.Detection.serialize` method, accordingly, since they all pass. Perfect!

Final note for today:
Perhaps [here](https://github.com/transientskp/pyse/blob/c56286ffcf99a88ebfcf15ca290e4636a86b4579/sourcefinder/extract.py#L1182) you should raise an AttributeError instead of a KeyError?

tmillenaar commented:
Good catch, changed to AttributeError.

I was able to reproduce your problems regarding `hatch run test:pytest`. It did work for me on other environments, like `hatch run default_310:pytest` . It took me quite some time but in the end I realized Hatch was just having caching problems and was using an older build for the test environment. Running `hatch env prune` resolved this for me. Can you verify if that also works in your case?

HannoSpreeuw commented:
Indeed `hatch env prune` fixed this.

Thanks for sorting that out, I had no clue.
